File: formal/coqimp/task05/extraction_ml/Zpower.ml
open BinNums
open BinPos
open Datatypes

(** val shift_nat : nat -> positive -> positive **)

let rec shift_nat n z =
  match n with
  | O -> z
  | S n0 -> Coq_xO (shift_nat n0 z)

(** val shift_pos : positive -> positive -> positive **)

let shift_pos n z =
  Pos.iter (fun x -> Coq_xO x) z n

(** val two_power_nat : nat -> coq_Z **)

let two_power_nat n =
  Zpos (shift_nat n Coq_xH)

File: formal/coqimp/task05/extraction_hs/Zpower.hs
module Zpower where

import qualified Prelude
import qualified BinPos
import qualified Datatypes

shift_nat :: Prelude.Integer -> Prelude.Integer -> Prelude.Integer
shift_nat n z =
  Datatypes.nat_rect z (\_ x -> (\x -> 2 Prelude.* x) x) n

shift_pos :: Prelude.Integer -> Prelude.Integer -> Prelude.Integer
shift_pos n z =
  BinPos._Pos__iter (\x -> (\x -> 2 Prelude.* x) x) z n

two_power_nat :: Prelude.Integer -> Prelude.Integer
two_power_nat n =
  (\x -> x) (shift_nat n 1)

